finality-aleph/src/lib.rs

```diff
@@ -89,7 +89,82 @@ pub struct MillisecsPerBlock(pub u64);
 #[derive(Copy, Clone, Debug, Default, Eq, PartialEq, Hash, Ord, PartialOrd, Encode, Decode)]
 pub struct UnitCreationDelay(pub u64);
 
-pub(crate) type SplitData<B> = Split<AlephNetworkData<B>, RmcNetworkData<B>>;
+pub type SplitData<B> = Split<AlephNetworkData<B>, RmcNetworkData<B>>;
+
+impl<B: Block> Versioned for AlephNetworkData<B> {
+    const VERSION: Version = Version(0);
+}
+
+#[derive(Encode, Eq, Decode, PartialEq)]
+pub struct Version(u32);
+
+pub trait Versioned {
+    const VERSION: Version;
+}
+
+/// It provides a generic implementation of the Decode and Encode traits by prepending byte
+/// representations for provided type parameters with their version (they need to implement the `Versioned` trait).
+/// If one provides data types with equal versions, the first data parameter will have priority while decoding.
+/// Keep in mind that with this approach, `decode` might fail even if the second data type could parse (fast fail).
+#[derive(Clone)]
+pub enum VersionedEitherMessage<L, R> {
+    Left(L),
+    Right(R),
+}
+
+impl<L: Versioned + Decode, R: Versioned + Decode> Decode for VersionedEitherMessage<L, R> {
```

Comment:
requiring Versioned here means we can only use this with types that we declared in finality-aleph (since we won't be able to implement this trait for foreign structs) or use some hacks like wrappers and so on. Could we just abandon this trait and use constants to denote versions for right/left variants of `VersionedEitherMessage`?

Reply:
You can implement local traits for foreign types without any additional wrappers. So should be able to implement it in finality-aleph for any Data type provided by aleph-bft. Or do you mean something different?